Add uri, urn and url test plugins

bcoca opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
tests/url.py

bcoca committed:
Add uri and url test plugins

bcoca committed:
added cripling loss of extra line

bcoca committed:
moar cases

bcoca committed:
moar cases, remove print

bcoca committed:
Another crissis averted via \n

bcoca committed:
added docs

bcoca committed:
aliasesd aliases

felixfontein commented on test/integration/targets/test_uri/tasks/main.yml:
```diff
@@ -0,0 +1,26 @@
+      - "'htp://google.com' is not url"   # bad schema
+      - "'ftp://google.com' is url"       # other schema
+      - "'file://etc/hosts' is not url"   # uris not urls
+      - "'mailto://me@nowhere.com' is not url"
```
I'm pretty sure that the two above are also URLs. (They have a protocol.)

bcoca replied:
see comment above

felixfontein commented on lib/ansible/plugins/test/uri.py:
```diff
@@ -0,0 +1,33 @@
+
+def is_url(value):
+    ''' Will verify that the string passed is a valid 'url' according to spec '''
+    return is_uri(value, ['http', 'ftp', 'https', 'ftps'])
```
Please link to the spec you mean here. The spec I know of, https://url.spec.whatwg.org/, already mentions several more valid schemes (see for example https://url.spec.whatwg.org/#url-miscellaneous), and if I understand it correctly basically any non-empty scheme string is allowed.

bcoca replied:
I mixed 'special scheme' with valid (also did not know websockets had been added to list, TIL), but mostly these are things we can deal with as urls as we have builtin handlers ... might need to give more options to user to handle 'unusable vs usable' in the context of urllib

bcoca replied:
I was going for mostly 'commonly usefule' but for full list of supported either using the urllib implementaiton
https://github.com/python/cpython/blob/3.10/Lib/urllib/parse.py#L47
or 'official list' https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml
but i'll just 'opt out' and leave it up to user

felixfontein commented on lib/ansible/plugins/test/uri.py:
```diff
@@ -0,0 +1,33 @@
+    ''' Will verify that the string passed is a valid 'uri' according to spec, if passed a list of valid schemas it will even match those '''
+    try:
+        x = urlparse(value)
+        isit = all([x.scheme, x.netloc, not schemas or x.scheme in schemas])
```
URIs do not need an non-empty netloc. That's only a requirement for URLs as far as I understand it.

bcoca replied:
yep, uri need x.path

felixfontein commented on test/integration/targets/test_uri/tasks/main.yml:
```diff
@@ -0,0 +1,26 @@
+      - "'asldkfjhalsidfthjo' is not uri" # junk
+      - "'http://admin:secret@example.com' is uri"
+      - "'ftps://admin:secret@example.com' is uri"
+      - "'admin:secret@example.com' is not uri"
```
This is a URI, but not a URL.

bcoca replied:
actually, depends on how you interpret `admin:` as it can be a schema/protocol, in this case we take it as part of auth and missing a schema (though you could argue that http is the implied schema, i was trying to avoid using implied)

If strictly using implied schemas  then `''` is a uri (implied http:// and empty path). But i'm open to suggestions on how to solve this quandary.

bcoca replied:
https://url.spec.whatwg.org/#urls
in this table `example` is both a valid and invalid URL, depending on implied base in this case. One reason i was trying to avoid implied parts as in the context of the test .. there is no other context (unless we allow passing that context in)

bcoca committed:
updated as per feedback

  test become a lot more permissive, might require user to
  pass in schemas to actually be useful

bcoca committed:
doc urn

bcoca committed:
updated tests and docs
t# Title: Summary, imperative, start upper case, don't end with a period

bcoca committed:
asldkfkj

Dule-martins commented on changelogs/fragments/added_uri_tests.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - new tests url and uri will verify a string matches the standard, but they don't check existance of the resource
```
```suggestion
  - URL and URI are tests that verify a string to match the standard, but they don't check the existence of resources.
```

Dule-martins commented on lib/ansible/plugins/test/uri.py:
```diff
@@ -0,0 +1,46 @@
+
+
+def is_uri(value, schemes=None):
+    ''' Will verify that the string passed is a valid 'uri', if passed a list of valid schemes it will match those '''
```
```suggestion
    ''' Will verify that the string passed is a valid 'URI', if given a list of valid schemes it will match those '''
```

Dule-martins commented on lib/ansible/plugins/test/uri.py:
```diff
@@ -0,0 +1,46 @@
+
+
+def is_url(value, schemes=None):
+    ''' Will verify that the string passed is a valid 'url' '''
```
```suggestion
    ''' Will verify that the string passed is a valid 'URL' '''
```

Dule-martins commented on lib/ansible/plugins/test/uri.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class TestModule(object):
+    ''' Ansible uri jinja2 tests '''
```
```suggestion
    ''' Ansible URI jinja2 test '''
```

Dule-martins commented on lib/ansible/plugins/test/uri.yml:
```diff
@@ -0,0 +1,30 @@
+  version_added: "2.14"
+  short_description: check if string is a valid URI
+  description:
+    - Validates that the input string conforms to the URI standard, optionally that is also is in the list of schemas provided.
```
```suggestion
    - Validates that the input string conforms to the URI standard, optionally that is also in the list of schemas provided.
```

Dule-martins commented on lib/ansible/plugins/test/urn.yml:
```diff
@@ -0,0 +1,21 @@
+  {{ 'mailto://nowone@example.com' is not urn }}
+RETURN:
+  _value:
+    description: Returns C(TRUE) if the string is a URN and C(False) if not
```
```suggestion
    description: Returns C(TRUE) if the string is a URN and C(False) if  it is not
```

Dule-martins reviewed (changes requested):
I did a bit of changes by adding and removing some few words from the file.

bcoca committed:
Update lib/ansible/plugins/test/uri.py

Co-authored-by: Tabah Baridule <dulemartins07@gmail.com>

bcoca committed:
Update lib/ansible/plugins/test/uri.py

Co-authored-by: Tabah Baridule <dulemartins07@gmail.com>

bcoca committed:
Update lib/ansible/plugins/test/uri.py

Co-authored-by: Tabah Baridule <dulemartins07@gmail.com>

bcoca committed:
Update lib/ansible/plugins/test/uri.yml

Co-authored-by: Tabah Baridule <dulemartins07@gmail.com>

bcoca committed:
Update lib/ansible/plugins/test/urn.yml

Co-authored-by: Tabah Baridule <dulemartins07@gmail.com>

bcoca committed:
upclog

samccann commented:
Docs portion LGTM

felixfontein commented on lib/ansible/plugins/test/uri.yml:
```diff
@@ -0,0 +1,30 @@
+    - Validates that the input string conforms to the URI standard, optionally that is also in the list of schemas provided.
+  options:
+    _input:
+      description: possible URI
```
```suggestion
      description: Possible URI.
```

felixfontein commented on lib/ansible/plugins/test/uri.yml:
```diff
@@ -0,0 +1,30 @@
+      type: string
+      required: True
+    schemes:
+      description: subset of URI schemas to validate against, otherwise ANY scheme is considered valid
```
```suggestion
      description: Subset of URI schemas to validate against, otherwise B(any) scheme is considered valid.
```

felixfontein commented on lib/ansible/plugins/test/uri.yml:
```diff
@@ -0,0 +1,30 @@
+  {{ 'http://nobody:secret@example.com' is uri(['ftp', 'ftps', 'http', 'https', 'ws', 'wss']) }}
+RETURN:
+  _value:
+    description: Returns C(False) if the string is not a URI or the schema extracted does not match the supplied list
```
```suggestion
    description: Returns C(false) if the string is not a URI or the schema extracted does not match the supplied list.
```

felixfontein commented on lib/ansible/plugins/test/url.yml:
```diff
@@ -0,0 +1,29 @@
+  version_added: "2.14"
+  short_description: check if string is a valid URL
+  description:
+    - Validates a string to conform to the URL standard
```
```suggestion
    - Validates a string to conform to the URL standard.
```

felixfontein commented on lib/ansible/plugins/test/url.yml:
```diff
@@ -0,0 +1,29 @@
+    - Validates a string to conform to the URL standard
+  options:
+    _input:
+      description: possible URL
```
```suggestion
      description: Possible URL.
```

felixfontein commented on lib/ansible/plugins/test/url.yml:
```diff
@@ -0,0 +1,29 @@
+      type: string
+      required: True
+    schemes:
+      description: subset of URI schemas to validate against, otherwise ANY scheme is considered valid
```
```suggestion
      description: Subset of URI schemas to validate against, otherwise B(any) scheme is considered valid.
```

felixfontein commented on lib/ansible/plugins/test/url.yml:
```diff
@@ -0,0 +1,29 @@
+  {{ 'ftp://admin:secret@example.com/path/to/myfile.yml' is url }}
+RETURN:
+  _value:
+    description: Returns C(False) if the string is not a URL, C(True) otherwise.
```
```suggestion
    description: Returns C(false) if the string is not a URL, C(true) otherwise.
```

felixfontein commented on lib/ansible/plugins/test/urn.yml:
```diff
@@ -0,0 +1,21 @@
+  version_added: "2.14"
+  short_description: check if string is a valid URN
+  description:
+    - Validates that the input string conforms to the URN standard
```
```suggestion
    - Validates that the input string conforms to the URN standard.
```

felixfontein commented on lib/ansible/plugins/test/urn.yml:
```diff
@@ -0,0 +1,21 @@
+    - Validates that the input string conforms to the URN standard
+  options:
+    _input:
+      description: possible URI
```
```suggestion
      description: Possible URN.
```

felixfontein commented on lib/ansible/plugins/test/urn.yml:
```diff
@@ -0,0 +1,21 @@
+  {{ 'mailto://nowone@example.com' is not urn }}
+RETURN:
+  _value:
+    description: Returns C(TRUE) if the string is a URN and C(False) if  it is not
```
```suggestion
    description: Returns C(true) if the string is a URN and C(false) if  it is not.
```

felixfontein commented on lib/ansible/plugins/test/uri.yml:
```diff
@@ -0,0 +1,30 @@
+      elements: string
+      required: False
+EXAMPLES: |
+  # urls are uris
```
```suggestion
  # URLs are URIs
```

felixfontein commented on lib/ansible/plugins/test/uri.yml:
```diff
@@ -0,0 +1,30 @@
+EXAMPLES: |
+  # urls are uris
+  {{ 'http://example.com' is uri }}
+  # but not all uris are urls
```
```suggestion
  # but not all URIs are URLs
```

felixfontein commented on lib/ansible/plugins/test/uri.yml:
```diff
@@ -0,0 +1,30 @@
+  {{ 'http://example.com' is uri }}
+  # but not all uris are urls
+  {{ 'mailto://nowone@example.com' is uri }}
+  # looking only for file transfers uris
```
```suggestion
  # looking only for file transfers URIs
```

felixfontein commented on lib/ansible/plugins/test/uri.yml:
```diff
@@ -0,0 +1,30 @@
+  {{ 'mailto://nowone@example.com' is uri }}
+  # looking only for file transfers uris
+  {{ 'mailto://nowone@example.com' is not uri(schemes=['ftp', 'ftps', 'sftp', 'file']) }}
+  # make sure url conforms to the 'special schemas'
```
```suggestion
  # make sure URL conforms to the 'special schemas'
```

felixfontein commented on lib/ansible/plugins/test/url.yml:
```diff
@@ -0,0 +1,29 @@
+      type: list
+      elements: string
+EXAMPLES: |
+  # simple url
```
```suggestion
  # simple URL
```

felixfontein commented on lib/ansible/plugins/test/url.yml:
```diff
@@ -0,0 +1,29 @@
+EXAMPLES: |
+  # simple url
+  {{ 'http://example.com' is url }}
+  # looking only for file transfers uris
```
```suggestion
  # looking only for file transfers URIs
```

felixfontein commented on lib/ansible/plugins/test/url.yml:
```diff
@@ -0,0 +1,29 @@
+  {{ 'mailto://nowone@example.com' is not uri(schemes=['ftp', 'ftps', 'sftp', 'file']) }}
+  #  but it is according to standard
+  {{ 'mailto://nowone@example.com' is not uri }}
+  # more complex url
```
```suggestion
  # more complex URL
```

felixfontein commented on lib/ansible/plugins/test/urn.yml:
```diff
@@ -0,0 +1,21 @@
+EXAMPLES: |
+  # urls are uris
+  {{ 'urn:isbn:9780302376463' is urn }}
+  # this is url/uri but not urn
```
```suggestion
  # this is URL/URI but not URN
```

felixfontein commented on lib/ansible/plugins/test/urn.yml:
```diff
@@ -0,0 +1,21 @@
+      type: string
+      required: True
+EXAMPLES: |
+  # urls are uris
```
```suggestion
  # ISBN in URN format
```

felixfontein commented on changelogs/fragments/added_uri_tests.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - new tests url, uri and urn will verify string as such, but they don't check existance of the resource
```
Very likely you won't need a changelog fragment for new test/filter plugins anymore, thanks to sidecar docs, but if you do, it should look as follows:
```suggestion
add plugin.test:
  - name: uri
    description: check if string is a valid URI
  - name: url
    description: check if string is a valid URL
  - name: urn
    description: check if string is a valid URN
```

felixfontein replied:
Actually no changelog fragment is needed anymore (just tried that), see also #77871.

bcoca committed:
Update lib/ansible/plugins/test/uri.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/uri.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/uri.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/url.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/url.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/url.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/urn.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/urn.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/url.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

felixfontein commented:
@bcoca there are 10 suggestions of mine hidden automatically by GH which you didn't respond to yet. Also I'm curious why you resolved the last comment without any comment or change.

Dule-mart commented:
@felixfontein 

The last come LGTM, what do you think?

felixfontein commented:
@Dule-mart I'm not sure what you mean.

bcoca commented:
@felixfontein i had merged most, ignored the changelog one as we still wanted it

bcoca committed:
Update lib/ansible/plugins/test/uri.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/uri.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/url.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/url.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/urn.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Update lib/ansible/plugins/test/urn.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Apply suggestions from code review

Co-authored-by: Felix Fontein <felix@fontein.de>
